gadget/gadget.go

```diff
@@ -551,6 +562,93 @@ type VolumeUpdate struct {
 	Preserve []string       `yaml:"preserve" json:"preserve"`
 }
 
+// VolumeAssignment is an optional set of volume-to-disk assignments
+// that can be specified. This is a nice way of reusing the same gadget
+// for multiple devices, or a way to map multiple volumes to the same disk
+// for cases like eMMC. Each assignment in this structure refers to a volume
+// and a device path (/dev/disk/** for now).
+type VolumeAssignment struct {
+	Name        string                       `yaml:"assignment-name"`
+	Assignments map[string]*DeviceAssignment `yaml:"assignment"`
+}
+
+func (va *VolumeAssignment) validate(volumes map[string]*Volume) error {
+	if len(va.Assignments) == 0 {
+		return fmt.Errorf("no assignments specified")
+	}
+
+	for name, asv := range va.Assignments {
+		if v := volumes[name]; v == nil {
+			return fmt.Errorf("volume %q is mentioned in assignment but has not been defined", name)
+		}
+		if err := asv.validate(); err != nil {
+			return fmt.Errorf("%q: %v", name, err)
+		}
+	}
+	return nil
+}
+
+// DeviceAssignment is the device data for each volume assignment. Currently
+// just keeps the device the volume should be assigned to.
+type DeviceAssignment struct {
+	Device string `yaml:"device"`
+}
+
+func (da *DeviceAssignment) validate() error {
+	// TODO: Revise this in the future to maybe limit it even further
```

Comment:
I am +1 with @bboozzoo on limiting this to by-path and by-id, and I actually think that we should limit now, not after people has started to use this. Moving in the other direction is much easier.